python/tk_hiero_export/sg_shot_processor.py

```diff
@@ -246,35 +300,283 @@ def startProcessing(self, exportItems):
+        cut_duration = 0
+
+        # this will hold a list of cut item data dicts that will be populated
+        # as the tasks are processed. these will be used to batch create the cut
+        # items at the end
+        cut_item_data_list = []
+
+        # process the tasks in order
+        for (shot_updater_task, transcode_task) in cut_related_tasks:
+
+            # cut order was populated by the calling method to update the
+            # Shot entity's cut info
+            cut_order = shot_updater_task._cut_order
+
+            # this retrieves the basic cut information from the updater task.
+            # cut item in/out, cut item duration, edit in/out.
+            cut_item_data = shot_updater_task.get_cut_item_data()
+
+            # add the length of this item to the full cut duration
+            cut_duration += cut_item_data["cut_item_duration"]
+
+            # translate some of the cut item data into timecodes that will also
+            # be populated in the cut item
+            tc_cut_item_in = self._timecode(cut_item_data["cut_item_in"], fps)
+            tc_cut_item_out = self._timecode(cut_item_data["cut_item_out"], fps)
+            tc_edit_in = self._timecode(cut_item_data["edit_in"], fps)
+            tc_edit_out = self._timecode(cut_item_data["edit_out"], fps)
+
+            # get the shot so that we have all we need for the cut item.
+            # this may create the shot if it doesn't exist already
+            shot = self.app.execute_hook(
+                "hook_get_shot",
+                task=shot_updater_task,
+                item=shot_updater_task._item,
+                data=self.app.preprocess_data,
+            )
+
+            # update the cut item data with the shot, timecodes and other fields
+            # required
+            cut_item_data.update({
+                "code": shot_updater_task.clipName(),
+                "project": self.app.context.project,
+                "shot": {"id": shot["id"], "type": "Shot"},
+                "cut_order": cut_order,
+                "timecode_cut_item_in": tc_cut_item_in,
+                "timecode_cut_item_out": tc_cut_item_out,
+                "timecode_edit_in": tc_edit_in,
+                "timecode_edit_out": tc_edit_out,
+            })
+
+            # add the populated cut item data to the list for batching later
+            cut_item_data_list.append(cut_item_data)
+
+            if cut_order == 1:
+                # first item in the cut, set the cut's start timecode
+                cut_data["timecode_start"] = tc_edit_in
+
+            if cut_order == len(cut_related_tasks):
+                # last item in the cut, set the cut's end timecode
+                cut_data["timecode_end"] = tc_edit_out
+
+        # all tasks processed, add the duration to the cut data
+        cut_data["duration"] = cut_duration
+
+        # create the cut to get the id.
+        sg = self.app.shotgun
+        cut = sg.create("Cut", cut_data)
```

Comment:
The thumbnail upload for the Cut was moved to the first shot updater task so as not to happen during pre-processing.